content/knowledge-base/guides/write-a-composition-function-in-python.md

```diff
@@ -0,0 +1,749 @@
+                    desired=fnv1beta1.State(
+                        resources={
+                            "xbuckets-test-bucket-a": fnv1beta1.Resource(
+                                resource=resource.dict_to_struct(
+                                    {
+                                        "apiVersion": "s3.aws.upbound.io/v1beta1",
+                                        "kind": "Bucket",
+                                        "metadata": {
+                                            "annotations": {
+                                                "crossplane.io/external-name": "test-bucket-a"
+                                            },
+                                        },
+                                        "spec": {
+                                            "forProvider": {"region": "us-east-2"}
+                                        },
+                                    }
+                                )
+                            ),
+                            "xbuckets-test-bucket-b": fnv1beta1.Resource(
+                                resource=resource.dict_to_struct(
+                                    {
+                                        "apiVersion": "s3.aws.upbound.io/v1beta1",
+                                        "kind": "Bucket",
+                                        "metadata": {
+                                            "annotations": {
+                                                "crossplane.io/external-name": "test-bucket-b"
+                                            },
+                                        },
+                                        "spec": {
+                                            "forProvider": {"region": "us-east-2"}
+                                        },
+                                    }
+                                )
+                            ),
+                        },
+                    ),
+                    context=structpb.Struct(),
+                ),
+            ),
+        ]
+
+        runner = fn.FunctionRunner()
+
+        for case in cases:
+            got = await runner.RunFunction(case.req, None)
+            self.assertEqual(
+                json_format.MessageToDict(got),
+                json_format.MessageToDict(case.want),
+                "-want, +got",
+            )
+
+
+if __name__ == "__main__":
+    unittest.main()
+```
+{{</expand>}}
+
+Run the unit tests using `hatch run`:
+
+```shell {copy-lines="1"}
+hatch run test:unit
+.
+----------------------------------------------------------------------
+Ran 1 test in 0.003s
+
+OK
+```
+
+{{<hint "tip">}}
+[Hatch](https://hatch.pypa.io/) is a Python build tool. It builds Python
+artifacts like wheels. It also manages virtual environments, similar
+to `virtualenv` or `venv`. The `hatch run` command creates a virtual environment
+and runs a command in that environment.
+
+You configure Hatch using `pyproject.toml`.
```

Comment:
```suggestion
```

Unless we're having them make changes I'd exclude this and leave it up to the Hatch docs to tell them where/what/how to customize.